flocker/provision/_digitalocean.py

```diff
@@ -0,0 +1,370 @@
+    """
+    # Import here, so that this can be added to ``flocker.provision`` without
+    # having to install ``pyocean``.
+    import pyocean
+
+    if (isinstance(exception, pyocean.exceptions.ClientError)
+        and exception.message == ('Droplet is currently on. '
+                                  'Please power it off to run this event.')):
+        return True
+    return False
+
+
+def retry_if_pending(callable, *args, **kwargs):
+    """
+    Repeats the API call if another API event is currently in progress.
+
+    It returns the result if the call eventually succeeds.
+
+    :param error_checkers: A ``list`` of ``callables`` which will check for
+        expected exceptions.
+    :param callable: The API function to call.
+    :param args: Positional arguments to supply when calling it.
+    :param kwargs: Keyword arguments to supply when calling it.
+    :return: The result of calling  ``callable``.
+    """
+    return retry_on_error([pending_event], callable, *args, **kwargs)
+
+
+def kernel_from_digitalocean_version(version):
+    """
+    Parse a DigitalOcean kernel version string into its component parts.
+
+    :param bytes version: The DigitalOcean kernel version string.
+    :return: ``Kernel`` with version information.
+    """
+    version, remaining = version.split('-', 1)
+    release, distribution, architecture = remaining.split('.', 2)
+    return Kernel(
+        version=version,
+        release=release,
+        distribution=distribution,
+        architecture=architecture
+    )
+
+
+DIGITAL_OCEAN_KERNEL_VERSION_TEMPLATE = (
+    '{version}-{release}.{distribution}.{architecture}'
+)
+
+
+def kernel_to_digitalocean_version(kernel):
+    """
+    Return a DigitalOcean style kernel string for the supplied ``Kernel``.
+    """
+    return DIGITAL_OCEAN_KERNEL_VERSION_TEMPLATE.format(
+        version=kernel.version,
+        release=kernel.release,
+        distribution=kernel.distribution,
+        architecture=kernel.architecture
+    )
+
+
+def set_droplet_kernel(droplet, required_kernel):
+    """
+    Change the kernel of the droplet with ``droplet_id``.
+
+    :param pyocean.Droplet droplet: The droplet whose kernel will be
+        configured.
+    :param Kernel required_kernel: The kernel version to be installed.
+    :returns: A ``pyocean.Kernel`` instance which was assigned to the droplet.
+    """
+    full_version = kernel_to_digitalocean_version(required_kernel)
+    for do_kernel in droplet.get_available_kernels():
+        if do_kernel.version == full_version:
+            break
+    else:
+        raise ValueError('Unknown kernel', required_kernel)
+
+    retry_if_pending(droplet.change_kernel, do_kernel.id)
+    return do_kernel
+
+
+def latest_droplet_kernel(droplet,
+                          required_distribution, required_architecture):
+    """
+    Return the newest kernel available for the supplied droplet, with the given
+    distribution and architecture.
+
```

Comment:
:question: Document raises. Marked optional because I'm not sure if this is something we have to do, but I have seen it around.